browser/ui/views/frame/brave_browser_view.cc

```diff
@@ -70,13 +80,30 @@
 
 namespace {
 absl::optional<bool> g_download_confirm_return_allow_for_testing;
+
+bool IsUnsupported(int command_id, Browser* browser) {
+  return chrome::IsRunningInForcedAppMode() &&
+         !chrome::IsCommandAllowedInAppMode(command_id,
+                                            browser->is_type_popup());
+}
+
 }  // namespace
 
 // static
 void BraveBrowserView::SetDownloadConfirmReturnForTesting(bool allow) {
   g_download_confirm_return_allow_for_testing = allow;
 }
 
+// static
+commands::Accelerators BraveBrowserWindow::GetDefaultAccelerators() {
```

Comment:
shouldn't this be in `brave_browser_window.cc`?
or did you intend this method to be `BraveBrowserView::GetDefaultAccelerators`?

Reply:
So this was to get around a deps violation - I wanted a static method on `BrowserWindow` implemented in the `views` layer, as it depends on `chrome/browser/ui/views/accelerator_table.h`

An instance method doesn't feel right, as there is only one set of default accelerators for all windows, it's not a per window type thing. Alternatively, I could also move the `AcceleratorService` & `AcceleratorServiceFactory` into `brave/browser/ui/views`. What do you think?

Reply:
I just tested moving it to `BraveBrowserWindow` and it triggers a deps violation, which is what I would've expected
![image](https://user-images.githubusercontent.com/7678024/222983311-acedef5b-1499-483b-8550-ed00a6bb8a67.png)

Doesn't this make sense as `chrome/browser/ui` is meant to be independent of the actual view implementation (which for desktop is in `chrome/browser/ui/views`)? (i.e. something in c/b/u/ shouldn't depend on c/b/u/views)?

Reply:
I believe current implementation is compliant with layering. I think @fallaciousreasoning was trying to solve layering violation where c/b/u depens on c/b/u/views with compile time polymorphism to make it clear that.

I think we can move this up other BraveBrowserView methods, and should have following code in brave_browser_window.cc

```
#if !defined(TOOLKIT_VIEWS)
commands::Accelerators BraveBrowserWindow::GetDefaultAccelerators() {
  NOTIMPLEMENTED();
  return {};
}
#endif

```

Reply:
> I just tested moving it to `BraveBrowserWindow` and it triggers a deps violation, which is what I would've expected ![image](https://user-images.githubusercontent.com/7678024/222983311-acedef5b-1499-483b-8550-ed00a6bb8a67.png)

that's a `DEPS` violation which is enforced with `DEPS` files. Usually when you see such failure (the one from `checkdeps.py` script), you need to:
1. Figure out if this dependency exist in `gn`. If not, make sure it's okay to add it and add it.
2. Reflect it in `DEPS` file.

[Let me fix our](https://github.com/brave/brave-core/pull/17481) `//brave/browser/ui` target first so things will just work in your PR. Currently it's pretty bad, so adding anything into `//brave/browser/ui` without required checks can break builds.

> Isn't `chrome/browser/ui` is meant to be independent of the actual view implementation (which for desktop is in `chrome/browser/ui/views`)? (i.e. something in c/b/u/ shouldn't depend on c/b/u/views)?

I understand you confusion, but this is not true currently. If you look closely at `//chrome/browser/ui` target, you will see that all UI views (`views/**` files) are declared there, not in `//chrome/browser/ui/views` target.

`//chrome/browser/ui/views` only declares a handful of files that are not actual views.

Reply:
> I understand you confusion, but this is not true currently. 

With all due respect, I think c/b/ui should be independent of UI tookits, even though currently there's only one toolkit we use(views//). For example, there're both c/b/ui/tabs and c/b/ui/views/tabs but files that depend on views// are under views/tabs.

> If you look closely at //chrome/browser/ui target, you will see that all UI views (views/** files) are declared there, not in //chrome/browser/ui/views target.

As far as I know, upstream folks have been trying to move views// specific targets into c/b/ui/views/, and views// specific things left in the c/b/ui/BUILD.gn is guarded at least.

---

Based on my understanding, I would say that moving this method to browser_window.cc isn't a good idea, as the method depends on views/accelerator_table.h.

---

cc. @simonhong , Could you tell us your take on this?

Reply:
> Based on my understanding, I would say that moving this method to browser_window.cc isn't a good idea, as the method depends on views/accelerator_table.h.

@sangwoo108 In this case what blocks us from having this method as `BraveBrowserView::GetDefaultAccelerators` in `brave_browser_view.cc`?

Reply:
Originally, I had the `accelerator_service` and `accelerator_service_factory` in `brave/browser/ui/views` for this reason (I had a similar understanding to @sangwoo108 about the separation of `brave/browser/ui/x` and `brave/browser/ui/views/x`). However, I think this service really does make more sense in `brave/browser/ui`. There's nothing about accelerators that are tied to the views framework (although it is the only place which uses them so far). `accelerator.h` lives in `ui/base/accelerators/accelerator.h` and so as a concept, they're independent of views.

We need to access the accelerator service outside of `views`, (from `//brave/browser/ui/webui` and possibly other places in future), and `webui` definitely shouldn't depend on `views` (as it could be displayed on Android, for example, although not in this case).

Before this PR, I was getting around this by adding a virtual method to `BrowserWindow` to expose the accelerators for that window - however, as the DefaultAccelerators are the same for all windows this felt a bit odd. I don't think it's that uncommon in Chromium to do it this way
`chrome/browser/ui/commander/commander_frontend.h::Create` is defined in `chrome/browser/ui/views/commander_frontend_views.cc` is the last example I came across.

Reply:
> I don't think it's that uncommon in Chromium to do it this way
> `chrome/browser/ui/commander/commander_frontend.h::Create` is defined in `chrome/browser/ui/views/commander_frontend_views.cc` is the last example I came across.

that's a good example of a clear code layout. Do you think it would be better to have a separate header in `ui` and a separate `..._views.cc` in `ui/views` for `commands::GetDefaultAccelerators()` instead of putting it into Window/View files?

Reply:
Cool, I've added a new `default_accelerators.h`/`default_accelerators_views.cc` file :smile: